[FLINK-2622][streaming]add WriteMode for writeAsText in DataStream.scala

HuangWHWHW opened the pull request:
Add a parameter writeMode in writeAsText with default setting "null".

tillrohrmann commented:
Thanks for your contribution @HuangWHWHW. However, your changes are not compatible to the Java DataStream API. Both APIs should support the same functionality with the same order of parameters. Furthermore, this PR is lacking tests.

HuangWHWHW commented:
@tillrohrmann 
Hi,
I see. 
The issue about order is just in the scala-example SocketTextStreamWordCount.scala as following:

{
if (fileOutput) {
      counts.writeAsText(outputPath, 1)
    } else {
      counts print
    }
}

So I changed the order for compile access.
As for the tests, I`ll try since I`m not good at scala programming.
BTW:Can you tell me why the CI is failed? It is still blocked in China.

tillrohrmann commented:
It's not only about the order but also about the support signatures. Best you take a look at the Java `DataStream` implementation to see what's supported.

The error is `error file=/home/travis/build/apache/flink/flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala message=File line length exceeds 100 characters line=721`. If you run `mvn clean verify` locally, you will spot these issues without Travis.

HuangWHWHW commented:
@tillrohrmann 
Hi,
Thank you for the CI info!
As I see, the writeAsText in scala DataStream calls the same method in Java DataStream(any misunderstanding?).
So the scala DataStream just pack the Java DataStream api.
Do you mean that there are some issues in other APIs(.e.g writeAsCsv())?

tillrohrmann commented:
I mean that the APIs are not consistent.

tillrohrmann commented:
In the Java DataStream API you can call the `writeAsText` method with:

* path
* path, write frequency
* path, writemode
* path. writemode, write frequency

In Scala this should also be supported.

HuangWHWHW commented:
Ah, I get you.
Maybe one writeAsText  method is not enough.
Add more writeAsText methods as JAVA DataStream?
And is there the same issue in writeAsCsv()?

tillrohrmann commented:
1. Yes
2. You should check

HuangWHWHW commented:
@tillrohrmann 
Thank you for the comments!
I will take the fix and add tests for this.
BTW:Can you take a look about this PR:https://github.com/apache/flink/pull/1030?
I got the CI passed and is there any new comment or could that be merged?
Very sorry for spending your time.

HuangWHWHW commented:
I take a fix for the consistent to JAVA and SCALA APIs.
As a result, we can call `writeAsText` and `writeAsCsv` method with:
*path
*path, write frequency
*path, writemode
*path. writemode, write frequency

And I add tests for these APIs.
And I noticed that in JAVA API, `writeAsCsv` cannot be configured the `Delimiter` for a row or a filed while the SCALA API is able.
I'm not sure neither to remove the feature in Scala API nor add it to the Java API.
So I added a new `writeAsCsv` method in Java DataStream for that we can configure the `Delimiter`.

HuangWHWHW commented:
@tillrohrmann 
Hi, I got a problem in this failed CI.
It is just a scala programming issue.
I found both in `DataStream.scala` and `DataSet.scala` had the same method now:
`def writeAsText(
      filePath: String,
      writeMode: FileSystem.WriteMode = null): DataSink[T]`
So in some failed tests(.e.g `org.apache.flink.api.scala.table.test.AsITCase`), `ds.writeAsText(resultPath, WriteMode.OVERWRITE)` cannot find the correct `writeAsText` method.
Could you tell me how to fix it?

HuangWHWHW commented:
I added the `javaSet` to the end of the table APIs.
I am not sure if it is suitable.

fhueske commented on flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java:
```diff
@@ -1004,10 +989,38 @@ public ExecutionConfig getExecutionConfig() {
+	 * @param writeMode
+	 *            Controls the behavior for existing files. Options are
+	 *            NO_OVERWRITE and OVERWRITE.
+	 * @param millis
```
Please add both delimiter parameters to the JavaDocs.

fhueske commented on flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:
```diff
@@ -727,13 +727,118 @@ class DataStream[T](javaStream: JavaStream[T]) {
+   * NO_OVERWRITE and OVERWRITE.
+   *
+   */
+  def writeAsText(
```
This method does not have a `millis` parameter. Please fix the documentation.

fhueske commented on flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:
```diff
@@ -727,13 +727,118 @@ class DataStream[T](javaStream: JavaStream[T]) {
    * every element of the DataStream the result of .toString
    * is written.
    *
+   * @param writeMode
```
please add the `path` parameter as well.

fhueske commented on flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:
```diff
@@ -727,13 +727,118 @@ class DataStream[T](javaStream: JavaStream[T]) {
+  }
+
+  /**
+   * Writes a DataStream to the file specified by path in text format. The
```
`writeToCsv` does not write `in text format` and does also not use the `.toString` method.
Please fix the documentation.

fhueske commented on flink-staging/flink-streaming/flink-streaming-scala/src/test/java/org/apache/flink/streaming/scala/api/CsvOutputFormatITCase.java:
```diff
@@ -36,6 +37,10 @@ protected void preSubmit() throws Exception {
 	@Override
 	protected void testProgram() throws Exception {
 		OutputFormatTestPrograms.wordCountToCsv(WordCountData.TEXT, resultPath);
+		OutputFormatTestPrograms.wordCountToCsv(WordCountData.TEXT, resultPath, 1);
```
Is this actually working? 
I thought by default files are not overwritten. So writing to the same file again, should not work and the second job should fail.

HuangWHWHW replied:
The `writeMode` by setting default is already exist.
It will check whether the file can be overwrite or not in `FileOutputFormat.java`:
`private static final void initDefaultsFromConfiguration(Configuration configuration) {
		final boolean overwrite = configuration.getBoolean(ConfigConstants
						.FILESYSTEM_DEFAULT_OVERWRITE_KEY,
				ConfigConstants.DEFAULT_FILESYSTEM_OVERWRITE);
	
		DEFAULT_WRITE_MODE = overwrite ? WriteMode.OVERWRITE : WriteMode.NO_OVERWRITE;
		
		final boolean alwaysCreateDirectory = configuration.getBoolean(ConfigConstants
						.FILESYSTEM_OUTPUT_ALWAYS_CREATE_DIRECTORY_KEY,
			ConfigConstants.DEFAULT_FILESYSTEM_ALWAYS_CREATE_DIRECTORY);
	
		DEFAULT_OUTPUT_DIRECTORY_MODE = alwaysCreateDirectory ? OutputDirectoryMode.ALWAYS : OutputDirectoryMode.PARONLY;
	}`

fhueske replied:
It is expected that the test fails if you set the overwrite mode to `NO_OVERWRITE` because you start multiple programs that all write to the same location.

I debugged the issue and found that the `ForkableFlinkMiniCluster` sets the overwrite mode to OVERWRITE. That explains the behavior. However, this also means that you cannot test the correctness of your methods by setting the overwrite mode to OVERWRITE because this is the default behavior.

I would change the `CsvOuputFormatITCase` to extend `StreamingMultipleProgramsTestBase` and run each program in a dedicated test method.

HuangWHWHW replied:
If it is expected to fail, could you tell me how to pass the test?
Just I have no idea about this.

fhueske commented:
Please check the documentation of the newly added functions.
- The text corresponds to the function (CSV and text output, not mentioning parameters which are not offered by the method, ...)
- All method parameters are described

Can you also double check that the test which overwrites files is working and if it is, why the write mode is set to OVERWRITE (should be NO_OVERWRITE)?
Thanks

HuangWHWHW commented:
@fhueske 
Hi, if I change to the NO_OVERWRITE in the `CsvOutputFormatITCase.java`, tests will fail with errors following:
`Caused by: java.io.IOException: File already exists:file:/C:/Users/H00292~1/AppData/Local/Temp/org.apache.flink.streaming.scala.api.CsvOutputFormatITCase-result/1
	at org.apache.flink.core.fs.local.LocalFileSystem.create(LocalFileSystem.java:247)
	at org.apache.flink.core.fs.local.LocalFileSystem.create(LocalFileSystem.java:263)
	at org.apache.flink.api.common.io.FileOutputFormat.open(FileOutputFormat.java:247)
	at org.apache.flink.api.scala.operators.ScalaCsvOutputFormat.open(ScalaCsvOutputFormat.java:161)
	at org.apache.flink.streaming.api.functions.sink.FileSinkFunction.open(FileSinkFunction.java:58)
	at org.apache.flink.api.common.functions.util.FunctionUtils.openFunction(FunctionUtils.java:33)
	at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.open(AbstractUdfStreamOperator.java:69)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.openAllOperators(StreamTask.java:232)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:167)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:581)
	at java.lang.Thread.run(Thread.java:722)`

This is due to the exist file check in `LocalFileSystem.java`.
See the code following:
`if (exists(f) && !overwrite) {
			throw new IOException("File already exists:" + f);
		}`

HuangWHWHW commented:
Hi, I think there is a issue that:
* if you want to no overwrite in a exist file, you will get the error "File already exists"
* if you want to no overwrite and the file is not exist, the `NO_OVERWRITE` is just the same as `OVERWRITE`
So I think currently the `NO_OVERWRITE` is not make sense.
Maybe we need discuss and fix on it through a new JIRA?

fhueske commented:
The semantics of `NO_OVERWRITE` are correct and pretty clear, IMO. 
If we would change `NO_OVERWRITE` to even fail if there is no file, it's semantics would change to `NO_WRITE` and always fail.

HuangWHWHW commented:
Doesn't `NO_OVERWRITE` mean "append"?

fhueske commented:
There is no append. `OVERWRITE` simply overwrites existing files and `NO_OVERWRITE` fails if the output file exists.

fhueske commented:
What's the purpose of adding `javaSet` to the Table API? 
This looks like a separate issue to me and should not be part of this PR if that is the case.

HuangWHWHW commented:
`What's the purpose of adding javaSet to the Table API? 
This looks like a separate issue to me and should not be part of this PR if that is the case.`

It is because of this following:
![image](https://cloud.githubusercontent.com/assets/13193847/9811803/11922a28-58ac-11e5-9f3f-4451f7db3b02.png)

chiwanpark commented:
I still cannot understand what is the purpose of adding `javaSet` to the Table API. We can get the `DataSet` by using `toDataSet` method.

chiwanpark commented:
Basically, we should prevent modifying the current Table API tests. This PR covers about streaming API only.

chiwanpark commented:
As you know, It seems because of ambiguousness between `Table` from `DataSet` and that from `DataStream`. We need refactor for Table API.

HuangWHWHW commented:
`I still cannot understand what is the purpose of adding javaSet to the Table API. We can get the DataSet by using toDataSet method.`

@fhueske 
@chiwanpark 
Hi, don't think too complicated.
This is just about the compile error for the Table API tests since I added the `def writeAsText( filePath: String, writeMode: FileSystem.WriteMode = null): DataSink[T]` method in `DataStream.scala`.
For easy understanding, you can also take a experiment that add the method and you will get some errors in these Table API tests.
To add the `javaSet` just a way that I can pass these tests and I'm not sure this change is suitable.
If not, how can I do?
Refactor Table API in a new JIRA or here?

HuangWHWHW commented:
@fhueske 
@chiwanpark 
Hello!
I opened a new PR:https://github.com/apache/flink/pull/1131 to fix the problem of Table API tests.
Is it valid?
I just want to pass the tests while fixing this PR.

fhueske commented:
Hi @HuangWHWHW, your PR #1131 has been merged.
Can you update this PR and remove those changes?
Thanks!

HuangWHWHW commented:
Yes, I will take a fix for the comments:)

HuangWHWHW commented:
I updated the PR:
1.Update the class and method documentation.
2.Change the test to `extends StreamingMultipleProgramsTestBase` and split the tests.
3.Add the fail tests for `NO_OVERWRITE`.

HuangWHWHW commented:
Hi, could anyone tell me why the last CI failed?
Thanks:)

chiwanpark commented:
There is a checkstyle error:

```
[INFO] There is 1 error reported by Checkstyle 6.2 with /tools/maven/checkstyle.xml ruleset.
[ERROR] src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java[992] (regexp) RegexpSinglelineJava: Line has leading space characters; indentation should be performed with tabs only.
```

HuangWHWHW commented:
Many thanks!
Fixed it.

tillrohrmann commented:
All checkstyle errors can also be found by simply executiong `mvn verify`
on your local machine. That way you can directly see what's going wrong.

On Sat, Sep 19, 2015 at 9:03 AM, HuangWHWHW <notifications@github.com>
wrote:

> Many thanks!
> Fixed it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/1098#issuecomment-141631769>.
>

HuangWHWHW commented:
@tillrohrmann 
Thanks!
For saving the time, I usually `run mvn` verify in the sub-module where I changed.
So, I forgot the DataStream.java in the other module before.

And I can see the CI detail now, and is the CI error following related to me?
`Tests in error: 
org.apache.flink.yarn.YARNSessionFIFOITCase.testJavaAPI(org.apache.flink.yarn.YARNSessionFIFOITCase)
  Run 1: YARNSessionFIFOITCase.testJavaAPI:642 » Runtime Unable to get Cluster status f...
  Run 2: YARNSessionFIFOITCase.checkForProhibitedLogContents:94->YarnTestBase.ensureNoProhibitedStringInLogFiles:294 Found a file /home/travis/build/apache/flink/flink-yarn-tests/target/flink-yarn-tests-fifo/flink-yarn-tests-fifo-logDir-nm-1_0/application_1442651724242_0009/container_1442651724242_0009_01_000001/jobmanager-main.log with a prohibited string: [Exception, Started SelectChannelConnector@0.0.0.0:8081]
`
I'll try to rerun the CI and may report this error if this is not due to my PR.

tillrohrmann commented:
This test shouldn't be related to your changes. Currently, our yarn tests
are a bit instable. But Robert is working on it.

On Mon, Sep 21, 2015 at 9:35 AM, HuangWHWHW <notifications@github.com>
wrote:

> @tillrohrmann <https://github.com/tillrohrmann>
> Thanks!
> For saving the time, I usually run mvn verify in the sub-module where I
> changed.
> So, I forgot the DataStream.java in the other module before.
>
> And I can see the CI detail now, and is the CI error following related to
> me?
> Tests in error:
>
> org.apache.flink.yarn.YARNSessionFIFOITCase.testJavaAPI(org.apache.flink.yarn.YARNSessionFIFOITCase)
> Run 1: YARNSessionFIFOITCase.testJavaAPI:642 » Runtime Unable to get
> Cluster status f...
> Run 2:
> YARNSessionFIFOITCase.checkForProhibitedLogContents:94->YarnTestBase.ensureNoProhibitedStringInLogFiles:294
> Found a file
> /home/travis/build/apache/flink/flink-yarn-tests/target/flink-yarn-tests-fifo/flink-yarn-tests-fifo-logDir-nm-1_0/application_1442651724242_0009/container_1442651724242_0009_01_000001/jobmanager-main.log
> with a prohibited string: [Exception, Started
> SelectChannelConnector@0.0.0.0:8081]
>
> I'll try to rerun the CI and may report this error if this is not due to
> my PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/1098#issuecomment-141898429>.
>

HuangWHWHW commented:
@tillrohrmann 
Thanks for the info.Passed the CI now.
Could anyone take a full review?
:)

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:
```diff
@@ -717,23 +717,188 @@ class DataStream[T](javaStream: JavaStream[T]) {
+      new Path(path),
+      ScalaCsvOutputFormat.DEFAULT_LINE_DELIMITER,
+      ScalaCsvOutputFormat.DEFAULT_FIELD_DELIMITER)
+    javaStream.write(of.asInstanceOf[OutputFormat[T]], millis)
```
Could you call `writeAsCsv(String, WriteMode, Long, String, String)` here with the respective parameter values instead? That way we have only one place where we have to apply changes.

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:
```diff
@@ -717,23 +717,188 @@ class DataStream[T](javaStream: JavaStream[T]) {
+    if (writeMode != null) {
+      of.setWriteMode(writeMode)
+    }
+    javaStream.write(of.asInstanceOf[OutputFormat[T]], 0L)
```
Same here: Call the most generic method here.

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/test/java/org/apache/flink/streaming/scala/api/TextOutputFormatITCase.java:
```diff
@@ -17,27 +17,86 @@
-	@Override
-	protected void testProgram() throws Exception {
+	@Test
+	public void failtestPathWriteModeMillis() throws Exception {
```
camel case

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/test/java/org/apache/flink/streaming/scala/api/CsvOutputFormatITCase.java:
```diff
@@ -17,32 +17,107 @@
+	public void failtestPathWriteModeMillisDelimiter() throws Exception {
+		OutputFormatTestPrograms.wordCountToCsv(WordCountData.TEXT, resultPath);
+		try {
+			OutputFormatTestPrograms.wordCountToCsv(WordCountData.TEXT, resultPath, FileSystem.WriteMode.NO_OVERWRITE, 1, "\n", ",");
```
Would be good to test the method with some other values than the default parameter values. That way you see that these values are correctly forwarded.

tillrohrmann commented on flink-staging/flink-streaming/flink-streaming-scala/src/test/java/org/apache/flink/streaming/scala/api/CsvOutputFormatITCase.java:
```diff
@@ -17,32 +17,107 @@
+	public void failtestPathWriteModeMillis() throws Exception {
+		OutputFormatTestPrograms.wordCountToCsv(WordCountData.TEXT, resultPath);
+		try {
+			OutputFormatTestPrograms.wordCountToCsv(WordCountData.TEXT, resultPath, FileSystem.WriteMode.NO_OVERWRITE, 1);
```
How do you test that the `millis` parameter is respected?

tillrohrmann commented:
Hi @HuangWHWHW, sorry for the late review. Things piled up in the last weeks. Do you wan to finish this PR?

I had some comments. I think that the PR is in a good shape right now. After addressing my comments it should be good to be merged.

tillrohrmann commented:
Apparently, you're no longer up for the task. I'll take over then.

fhueske commented:
Thanks for working on this issue!
I'll close this PR due to inactivity and PR #1473.
